plugins/inputs/tencentcloudcm/tencentcloudcm.go

```diff
@@ -0,0 +1,375 @@
+					return fmt.Errorf("region is empty")
+				}
+				_, ok := t.discoverTool.registry[namespace.Name]
+				if len(region.Instances) == 0 && !ok {
+					return fmt.Errorf(
+						"unsupported namespace %s for discovering instances, please specify instances and dimensions",
+						namespace.Name,
+					)
+				}
+			}
+		}
+	}
+
+	// start discovery
+	go t.discoverTool.Discover(t.Accounts, t.DiscoveryInterval, t.Endpoint)
+	t.discoverTool.DiscoverMetrics()
+	for i, account := range t.Accounts {
+		for j, namespace := range account.Namespaces {
+			// use discovered metrics if not specified
+			if len(namespace.Metrics) == 0 {
+				metrics, ok := t.discoverTool.DiscoveredMetrics[namespace.Name]
+				if !ok {
+					return fmt.Errorf("unsupported namespace %s for discovering metrics, please specify metrics",
+						namespace.Name)
+				}
+				t.Accounts[i].Namespaces[j].Metrics = metrics
+			}
+		}
+	}
+
+	t.client = &cloudmonitorClient{Accounts: t.Accounts, Log: t.Log}
+
+	return nil
+
+}
+
+func (t *TencentCloudCM) updateWindow(relativeTo time.Time) {
+	windowEnd := relativeTo.Add(-time.Duration(t.Delay))
+	// starting point is two times the aggregation period to make sure all points are covered
+	t.windowStart = windowEnd.Add(-time.Duration(t.Period) * 2)
+	t.windowEnd = windowEnd
+}
+
+func (t *TencentCloudCM) Gather(acc telegraf.Accumulator) error {
+	t.updateWindow(time.Now())
+
+	metricObjects := t.client.GetMetricObjects(*t)
+
+	lmtr := limiter.NewRateLimiter(t.RateLimit, time.Second)
+	defer lmtr.Stop()
+
+	wg := sync.WaitGroup{}
+	rLock := sync.Mutex{}
+	results := []monitor.GetMonitorDataResponse{}
+
+	// requestIDMap contains request ID and metric objects for later aggregation
+	requestIDMap := map[string]metricObject{}
+	for _, obj := range metricObjects {
+		wg.Add(1)
+		<-lmtr.C
+		go func(m metricObject) {
+			defer wg.Done()
+
+			for {
+
+				client, err := t.client.NewClient(m.Region, m.Account.crs, *t)
+				if err != nil {
+					acc.AddError(err)
+					return
+				}
+
+				if len(m.MonitorInstances) >= t.BatchSize {
+					batch := m.MonitorInstances[:t.BatchSize]
+					if len(batch) == 0 {
+						break
+					}
+
+					request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, batch, *t)
+
+					result, err := t.client.GatherMetrics(client, request, *t)
+					if err != nil {
+						acc.AddError(err)
+						break
+					}
+
+					rLock.Lock()
+					requestIDMap[*result.Response.RequestId] = m
+					results = append(results, *result)
+					rLock.Unlock()
+
+					m.MonitorInstances = m.MonitorInstances[t.BatchSize:]
+				} else {
+					request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, m.MonitorInstances, *t)
+
+					result, err := t.client.GatherMetrics(client, request, *t)
+					if err != nil {
+						acc.AddError(err)
+						break
+					}
+
+					rLock.Lock()
+					requestIDMap[*result.Response.RequestId] = m
+					results = append(results, *result)
+					rLock.Unlock()
+					break
+				}
```

Comment:
Can this be simplified a bit, e.g.
```suggestion
				batch := m.MonitorInstances
				if len(m.MonitorInstances) >= t.BatchSize {
					batch := m.MonitorInstances[:t.BatchSize]
				}
				if len(batch) == 0 {
					break
				}

				request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, batch, *t)

				result, err := t.client.GatherMetrics(client, request, *t)
				if err != nil {
					acc.AddError(err)
					break
				}

				rLock.Lock()
				requestIDMap[*result.Response.RequestId] = m
				results = append(results, *result)
				rLock.Unlock()

				m.MonitorInstances = m.MonitorInstances[len(batch):]
```

Reply:
refactored in commit 31d8e6c